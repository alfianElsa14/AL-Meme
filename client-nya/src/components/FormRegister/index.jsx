import React, { useState } from 'react'
import { FormattedMessage } from 'react-intl';
import { useNavigate } from 'react-router-dom'
import { useDispatch } from 'react-redux'

import { registerUser } from '@containers/Client/actions'

import classes from './style.module.scss'

function FormRegister() {
    const navigate = useNavigate()
    const dispatch = useDispatch()

    const [username, setUsername] = useState('');
    const [email, setEmail] = useState('');
    const [password, setPassword] = useState('');
    const [imageUrl, setImageUrl] = useState(null);
    const [tempImage, setTempImage] = useState(null)

    const handleImageChange = (e) => {
        const file = e.target.files[0];

        if (file) {
            const gambar = URL.createObjectURL(file);
            setImageUrl(file);
            setTempImage(gambar)
        } else {
            setImageUrl(null);
        }
    }

    const handleSubmit = (e) => {
        e.preventDefault();

        const formData = new FormData();
        formData.append('username', username);
        formData.append('email', email);
        formData.append('password', password);

        if (imageUrl) {
            formData.append('imageUrl', imageUrl);
        }
        


        dispatch(registerUser(formData, navigate));
    }

    return (
        <div
            className={classes.formContainer}
            data-testid="formContainer"
        >
            <h3><FormattedMessage id="app_register" /></h3>
            <div className={classes.form}>
                <form
                    data-testid="formRegister"
                    onSubmit={handleSubmit} action="" className={classes.formRegister}>
                    <div className={classes.picture}>
                        <img src={tempImage || 'https://static.vecteezy.com/system/resources/previews/020/765/399/non_2x/default-profile-account-unknown-icon-black-silhouette-free-vector.jpg'} alt="" />
                        <label htmlFor="image" className={classes.customFileButton}>
                            <button>
                                <FormattedMessage id="app_change" />
                            </button>
                            <input type="file" id="image"
                                onChange={handleImageChange} name="imageUrl" className={classes.fileInput} />
                        </label>
                    </div>
                    <div className={classes.inputItem}>
                        <label htmlFor="username">
                            <FormattedMessage id="app_username" /> :
                        </label>
                        <input type="text" id="username" name="username"
                            value={username}
                            onChange={(e) => setUsername(e.target.value)} 
                            data-testid="username"
                            />
                    </div>
                    <div className={classes.inputItem}>
                        <label htmlFor="email">
                            Email :
                        </label>
                        <input
                            type="email"
                            id="email"
                            name="email"
                            value={email}
                            onChange={(e) => setEmail(e.target.value)}
                            data-testid="email"
                        />
                    </div>
                    <div className={classes.inputItem}>
                        <label htmlFor="password">
                            Password :
                        </label>
                        <input
                            type="password"
                            id="password"
                            name="password"
                            value={password}
                            onChange={(e) => setPassword(e.target.value)}
                            data-testid="password"
                        />
                    </div>
                    <button type="submit">
                        <FormattedMessage id="app_register" />
                    </button>
                </form>
            </div>
            <p>
                <FormattedMessage id="app_already_have_an_account" /> click <span 
                data-testid="toLogin"
                onClick={() => navigate('/login')}>
                    <FormattedMessage id="app_here" />
                </span>
            </p>
        </div>
    )
}

export default FormRegister