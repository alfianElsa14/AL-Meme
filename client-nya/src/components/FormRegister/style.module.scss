@import '@styles/base';

.formContainer {
    width: 100%;
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    border-left: none;
    color: $color-white;
    background-color: $color-marine-blue;
    border-radius: 5px;

    @include apply-at-md-size {
        padding: 20px;
    }

    .formRegister {
        display: flex;
        flex-direction: column;
        width: 300px;
        height: 100%;
        color: $color-white;

        .picture {
            display: flex;
            flex-direction: column;
            justify-content: center;
            align-items: center;

            .img {
                width: 60px;
                height: 60px;
                object-fit: cover;
                border-radius: 50%;
                border: none;
            }

            .customFileButton {
                position: relative;
                overflow: hidden;
                display: inline-block;

                & button {
                    display: flex;
                    justify-content: center;
                    align-items: center;
                    border: none;
                    background-color: $color-purplish-blue;
                    color: $color-white;
                    height: 30px;
                    margin-top: 15px;
                    margin-bottom: 15px;
                    border-radius: 3px;
                    width: 100px;
                    font-size: 15px;
                    cursor: pointer;

                    &:hover {
                        background-color: rgb(100, 100, 218);
                    }
                }

                & input[type="file"] {
                    font-size: 100px;
                    position: absolute;
                    left: 0;
                    top: 0;
                    opacity: 0;
                    cursor: pointer;
                }
            }
        }

        .inputItem {
            @include directionY;

            label {
                font-size: 0.6rem;
                display: flex;
                justify-content: space-between;
            }

            .error {
                color: red;
                font-size: 0.5rem;
                font-weight: 400;
            }

            input {
                padding: 0.3rem;
                font-size: 0.6rem;
                margin-bottom: 1rem;
                border-radius: 5px;
                background-color: $color-white;

                @include apply-at-md-size {
                    padding: 0.3rem;
                    font-size: 0.6rem;
                    margin-bottom: 1rem;
                    border-radius: 5px;
                    background-color: $color-white;
                }
            }

            select {
                padding: 0.3rem;
                font-size: 0.6rem;
                margin-bottom: 1rem;
                border-radius: 5px;
                background-color: rgb(211, 211, 211);
                cursor: pointer;

                option {
                    cursor: pointer;
                }
            }
        }

        button {
            border: none;
            height: 40px;
            border-radius: 5px;
            background-color: $color-light-blue;
            cursor: pointer;
        }

        button:hover {
            background-color: rgb(0, 105, 203);
            color: $color-aqua;
        }
    }

    p {
        font-size: 15px;

        a:hover {
            color: $color-aqua;
        }
    }
}

p {
    span {
        font-style: italic;
        color: blue;
        cursor: pointer;
        text-decoration: underline;
    }

    span:hover {
        color: rgb(167, 167, 255);
    }
}