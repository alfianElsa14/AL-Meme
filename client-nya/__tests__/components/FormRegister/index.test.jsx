import { fireEvent, render } from "@utils/testHelper";
import classes from "@components/FormRegister/style.module.scss"
import FormRegister from "@components/FormRegister";

const mockNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
    useNavigate: () => mockNavigate,
}));

let wrapper;

beforeEach(() => {
    wrapper = render(<FormRegister />)
})

describe("Form Register", () => {
    test('Should render correctly', () => {
        const { getByTestId } = wrapper;
        const FormRegisterComponent = getByTestId('formContainer');
        expect(FormRegisterComponent).toHaveClass(classes.formContainer)
        expect(FormRegisterComponent).toBeInTheDocument();
    })

    test('Should render form register correctly', () => {
        const { getByTestId } = wrapper;
        const FormRegister = getByTestId('formRegister');
        expect(FormRegister).toHaveClass(classes.formRegister)
        expect(FormRegister).toBeInTheDocument();
    })

    test('Should handle input username change', () => {
        const { getByTestId } = wrapper;
        const emailInput = getByTestId('username');
        fireEvent.change(emailInput, { target: { value : 'luak'}})
        expect(emailInput.value).toBe('luak')
    })

    test('Should handle input email change', () => {
        const { getByTestId } = wrapper;
        const emailInput = getByTestId('email');
        fireEvent.change(emailInput, { target: { value : 'luak@gmail.com'}})
        expect(emailInput.value).toBe('luak@gmail.com')
    })

    test('Should handle input password change', () => {
        const { getByTestId } = wrapper;
        const passwordInput = getByTestId('password');
        fireEvent.change(passwordInput, { target: { value : '1234567'}})
        expect(passwordInput.value).toBe('1234567')
    })

    test('Should call navigate when text clicked', () => {
        const { getByTestId } = wrapper;
        const text = getByTestId('toLogin');
        fireEvent.click(text);
        expect(mockNavigate).toHaveBeenCalledWith(`/login`);
    })

    test('Should match with snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    })
})