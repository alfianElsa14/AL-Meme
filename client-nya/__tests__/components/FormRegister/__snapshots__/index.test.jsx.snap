// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form Register Should match with snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="formContainer"
        data-testid="formContainer"
      >
        <h3>
          Daftar
        </h3>
        <div
          class="form"
        >
          <form
            action=""
            class="formRegister"
            data-testid="formRegister"
          >
            <div
              class="picture"
            >
              <div
                class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault img css-2s90m6-MuiAvatar-root"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiAvatar-fallback css-10mi8st-MuiSvgIcon-root-MuiAvatar-fallback"
                  data-testid="PersonIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
                  />
                </svg>
              </div>
              <label
                class="customFileButton"
                for="image"
              >
                <button>
                  Ganti
                </button>
                <input
                  class="fileInput"
                  id="image"
                  name="imageUrl"
                  type="file"
                />
              </label>
            </div>
            <div
              class="inputItem"
            >
              <label
                for="username"
              >
                Nama panggilan
                 :
              </label>
              <input
                data-testid="username"
                id="username"
                name="username"
                type="text"
                value=""
              />
            </div>
            <div
              class="inputItem"
            >
              <label
                for="email"
              >
                Email :
              </label>
              <input
                data-testid="email"
                id="email"
                name="email"
                type="email"
                value=""
              />
            </div>
            <div
              class="inputItem"
            >
              <label
                for="password"
              >
                Password :
              </label>
              <input
                data-testid="password"
                id="password"
                name="password"
                type="password"
                value=""
              />
            </div>
            <button
              type="submit"
            >
              Daftar
            </button>
          </form>
        </div>
        <p>
          Sudah punya akun?
           click 
          <span
            data-testid="toLogin"
          >
            Disini
          </span>
        </p>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="formContainer"
      data-testid="formContainer"
    >
      <h3>
        Daftar
      </h3>
      <div
        class="form"
      >
        <form
          action=""
          class="formRegister"
          data-testid="formRegister"
        >
          <div
            class="picture"
          >
            <div
              class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault img css-2s90m6-MuiAvatar-root"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiAvatar-fallback css-10mi8st-MuiSvgIcon-root-MuiAvatar-fallback"
                data-testid="PersonIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
                />
              </svg>
            </div>
            <label
              class="customFileButton"
              for="image"
            >
              <button>
                Ganti
              </button>
              <input
                class="fileInput"
                id="image"
                name="imageUrl"
                type="file"
              />
            </label>
          </div>
          <div
            class="inputItem"
          >
            <label
              for="username"
            >
              Nama panggilan
               :
            </label>
            <input
              data-testid="username"
              id="username"
              name="username"
              type="text"
              value=""
            />
          </div>
          <div
            class="inputItem"
          >
            <label
              for="email"
            >
              Email :
            </label>
            <input
              data-testid="email"
              id="email"
              name="email"
              type="email"
              value=""
            />
          </div>
          <div
            class="inputItem"
          >
            <label
              for="password"
            >
              Password :
            </label>
            <input
              data-testid="password"
              id="password"
              name="password"
              type="password"
              value=""
            />
          </div>
          <button
            type="submit"
          >
            Daftar
          </button>
        </form>
      </div>
      <p>
        Sudah punya akun?
         click 
        <span
          data-testid="toLogin"
        >
          Disini
        </span>
      </p>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
